// framework/Util.ts
const path = require("path");
const fs = require("fs");

export function MakeUtil() {
    return new Util();
}

export class Util {

    get uploadDirectory() {
        return path.join(__dirname, "../../../upload/");
    }

    get tmpDirectory() {
        return path.join(__dirname, "../../../tmp/");
    }

}

export function isEmptyStr(str: string) {
    return str && str.trim().length <= 0;
}

export function getFilename(path_: string) {
    const tmp = path_ ? path_.split("/") : [""];
    return tmp[tmp.length - 1];
}

export function parseUrl(url: string): string {
    return url ? url.replace(/"/g, "") : "";
}

export function isNotNullNorUndefined(object: any) {
    return object !== null && object !== undefined;
}

export function getFullName(firstName: string, lastName: string) {
    if (!firstName && !lastName) {
            return "Anonymous";
        } else if (!lastName) {
            return firstName;
        } else if (!firstName) {
            return lastName;
        } else {
            return firstName + " " + lastName;
        }
}

export function readJSON(path) {
    const json = fs.readFileSync(path);
    return JSON.parse(json);
}

export function resourcePath() {
    return path.join(__dirname, "../../../resources");
}
