// domain/infoAllPage/InfoAllPage.ts
import {Entity} from "domain/Entity";
import {SequelizeClass} from "application/SequelizeClass";
const Connection = SequelizeClass.getConnection().connection;
const Sequelize = SequelizeClass.getConnection().Sequelize;

export interface  CreateInfoAllPage {
    dataInfo: JSON;
}

export interface InfoAllPage extends Entity {
    getId(): string;
}

export function InfoAllPageRecordFrom (record: any): InfoAllPage {
    return new InfoAllPageImpl (record);
}

export class InfoAllPageImpl implements InfoAllPage {
    private _record : any;

    constructor(record: any){
        this._record = record;
        Connection.sync();
    }
    get record(): any{
        return this._record;
    }

    public static SCHEMA = Connection.define('InfoAllPage', {
        id:{type: Sequelize.STRING, unique: true, allowNull: false, primaryKey: true},
        dataInfo: {type: Sequelize.JSON}
    },{
        freezeTableName: true,
        paranoid:true
    });

    public toJSON(){
        return {
            id: this.record.id,
            dataInfo: this.record.dataInfo
        }
    }

    public getId(){
        return this.record.id;
    }
}
